.github/scripts/ensure_actions_will_cancel.py

```diff
@@ -0,0 +1,126 @@
+    "update_s3_htmls.yml",
+    "push_nightly_docker_ghcr.yml",
+}
+
+
+def should_skip(filename):
+    if filename.name in SKIP:
+        return True
+
+    with open(filename, "r") as f:
+        content = f.read()
+        if "@generated DO NOT EDIT MANUALLY" in content:
+            return True
+
+    data = yaml.safe_load(content)
+    on = str(data.get("on", data.get(True, None)))
+    if "'pull_request'" not in on and "'workflow_dispatch'" not in on:
+        return True
+
+    return False
+
+
+def insert_concurrency_keys(filename):
+    # Import here since this isn't installed in CI where this script runs
+    from ruamel.yaml import YAML
+
+    yaml = YAML()
+    yaml.indent(mapping=2, sequence=4, offset=2)
+    yaml.preserve_quotes = True
+    yaml.width = 1000
+    data = yaml.load(filename)
+
+    on = str(data["on"])
+    if "'pull_request'" not in on and "'workflow_dispatch'" not in on:
+        print(f"Skipping non-PR workflow '{filename}'")
+        return
+
+    data["concurrency"] = {
+        "group": concurrency_key(filename),
+        "cancel-in-progress": True,
+    }
+
+    buf = io.StringIO()
+    yaml.dump(data, buf)
+
+    # ruamel.YAML doesn't put in a newline so do that to make it look nicer
+    content = buf.getvalue()
+    content = re.sub(
+        r"\nconcurrency:\n",
+        r"\n\nconcurrency:\n",
+        content,
+        flags=re.MULTILINE,
+    )
+
+    with open(filename, "w") as f:
+        f.write(content)
+    print(f"Wrote {filename}")
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Ensure all relevant GitHub actions jobs will be cancelled based on a concurrency key"
+    )
+    parser.add_argument(
+        "--write",
+        action="store_true",
+        default=False,
+        required=False,
+        help="insert concurrency keys instead of checking",
+    )
+    args = parser.parse_args()
+
+    files = WORKFLOWS.glob("*.yml")
+
+    errors_found = False
+    files = [f for f in files if not should_skip(f)]
+    for filename in files:
+        if args.write:
+            # Go and try to insert the relevant entry for the concurrency group
+            insert_concurrency_keys(filename)
+        else:
+            with open(filename, "r") as f:
+                data = yaml.safe_load(f)
+
+            expected = {
+                "group": concurrency_key(filename),
+                "cancel-in-progress": True,
+            }
+            if "concurrency" not in data.keys():
+                print(
+                    f"'concurrency' not found in '{filename.relative_to(REPO_ROOT)}'",
+                    file=sys.stderr,
+                )
+                errors_found = True
+            elif data["concurrency"] != expected:
+                print(
+                    f"'concurrency' incorrect in '{filename.relative_to(REPO_ROOT)}'",
+                    file=sys.stderr,
+                )
+                errors_found = True
```

Comment:
it seems unfortunate for this to be a separate codepath; what are your thoughts on this alternate approach?
1. regardless of whether `--write` is enabled, first read the file (and save its string form for later)
2. parse the YAML and and set its `concurrency` key (as you already do in `insert_concurrency_keys`
3. insert a newline and whatever else you need to do to make it nice
4. if `--write` is enabled, just write that back to the file
5. otherwise, compare it with what you read earlier

Reply:
alternatively, you could eliminate the `--write` parameter altogether, and then just use `git diff --exit-code` like we were doing before, which is better UX because then it tells the PR author exactly what they need to do to fix the issue

Reply:
I'm weary of that since it will also essentially be a yaml style linter since round-tripping from yaml -> Python -> yaml doesn't always work great if you aren't being super consistent with your whitespace / quotes / etc (ie. see the whitespaces changes in this PR). Here the `--write` is just a helper that's in the same file for convenience's sake mostly, I'd be fine taking it out and making this only do the checking since we probably won't need to auto-add a bunch of `concurrency:` stuff again

Reply:
yeah that's fair